### DISCUSSION TITLE
Détails techniques du routeur Tri mk2...

### DISCUSSION THREAD

**Discussion by cvfe13** [FredM67/PVRouter-3-phase#90] [General]
Bonjour Fred,

Je suis en train de monter le routeur tri acquis auprès de mk2PVRouter.com et plusieurs points ne sont pas très clairs pour moi.

### 1. Concernant les sondes de température Dallas:

> Il est possible de brancher un ou plusieurs capteurs de température Dallas DS18B20.
> Ces capteurs peuvent servir à des fins informatives ou pour contrôler le mode de fonctionnement forcé.

Comment lire les valeurs de ces sondes ?

### 2. Rotation des priorités
Lorsqu'on choisit l'option manuelle de rotation des priorités, je suppose que le signal sur la PIN choisie est une impulsion ?
Doit-on appliquer un état haut ou un état bas ?
Faut-il une résistance de pull-up ?

### 4. EMONESP
Ces deux dernières fonctions sont d'ailleurs activées par defaut si l'on utilise EMONESP. Je ne connais pas EMONESP, mais je suppose que c'est un moyen de commander ces fonctions via un ESP8266. Alors comment raccorder cetr ESP au routeur ? Peut-on lire les valeurs de routage de cette façon ?

### 5. Priorité des sorties relais
Est-il possible de commuter une sortie relais pendant la phase de routage? Je m'explique:
il me faut en moyenne 3kWh d'energie pour 'recharger' mon ECS. Lorsque ma production solaire est importante, l'ECS sera à sa température maximale dès 10:00 du matin. Il n'y a qu'en hiver que je peux jouer sur une ou deux charges résistives absorber le surplus.
En été, les clims ou pompes de piscine commandées en Tout ou Rien permettraient 'd'étaler' le routage sur une période de la journée beaucoup plus longue.

### 6. Calibration..
Je vois:

> f_powerCal{ 0.04320F, 0.04374F, 0.04385F };
> f_voltageCal{ 0.7892F, 0.8002F, 0.7995F }; 

Faut-il calibrer également la tension ?

### 7. Offset de routage
Y a t'il un paramètre d'offset pour dire au routeur de rechercher par exemple 20W d'import (ou d'export)

Francis

**Comment by FredM67** [comment]
Bonjour Francis,

### 1. Concernant les sondes de température Dallas:
> > Il est possible de brancher un ou plusieurs capteurs de température Dallas DS18B20.
> > Ces capteurs peuvent servir à des fins informatives ou pour contrôler le mode de fonctionnement forcé.
>
> Comment lire les valeurs de ces sondes ?

Dans le programme, la lecture des sondes se fait de facon asynchrone en raison du délai (énorme) d'environ 750ms.
Il faut d'une part faire une demande `requestTemperatures` ensuite appeler `readTemperature`.

### 2. Rotation des priorités
> Lorsqu'on choisit l'option manuelle de rotation des priorités, je suppose que le signal sur la PIN choisie est une impulsion ? Doit-on appliquer un état haut ou un état bas ? Faut-il une résistance de pull-up ?

Par défaut, et sauf mention contraire dans le prog, les pin configurées en entrée sont en application d'un état bas.
Pour la rotation, il s'agit d'une "impulsion" d'au moins une seconde pour être sûr que l'état soit lu. Si cela pose problème, il est possible de changer pour accepter des impulsions plus courtes.
Non, pas de résistance pull-up, on utilise la pull-up interne de l'Arduino.

### 3. Idem pour la Marche forcée et l'arrêt du routage

Ici, c'est différent. La durée de la marche forcée et de l'arrêt du routage sont déterminées par la durée de l'état bas appliqué.
Ainsi, si je veux une marche forcée pendant 2h, j'applique pendant 2h un état bas. J'ai fait cela comme ca car c'est bcp plus simple à gérer en interne, notamment en raison de l'absence de sauvegarde des états/données en cas de redémarrage/coupure de courant/...
Cette méthode est également compatible avec une horloge programmable par exemple.

### 4. EMONESP
> Ces deux dernières fonctions sont d'ailleurs activées par defaut si l'on utilise EMONESP. Je ne connais pas EMONESP, mais je suppose que c'est un moyen de commander ces fonctions via un ESP8266. Alors comment raccorder cetr ESP au routeur ? Peut-on re les valeurs de routage de cette façon ?

Dans le code, il est question d'EmonESP car c'est le module qui m'a servi d'inspiration. Il est basé sur un ESP8266 ou ESP32.. mais bon, n'importe quoi peu faire l'affaire, tant que ce module fournit des sorties type contact sec ou compatibles avec les tensions de l'Arduino.
Pour le raccorder, il faut d'une part connecter les interfaces série de l'EmonESP et de l'Arduino, et d'autre part les pin avec GPIO correspondantes pour les fonctions marche forcée, arrêt routage, ...
On peut lire les valeurs disponibles dans le routeur, à savoir le bilan de puissance sur chaque phase et global, les tensions de chaque phase et les température de chaque capteur si présents.
Accessoirement, certains se sont amusés à lire de taux de routage par chaque charge (charge alimentée à 25%, 30%, ...).

### 5. Priorité des sorties relais
Est-il possible de commuter une sortie relais pendant la phase de routage? Je m'explique: il me faut en moyenne 3kWh d'energie pour 'recharger' mon ECS. Lorsque ma production solaire est importante, l'ECS sera à sa température maximale dès 10:00 du tin. Il n'y a qu'en hiver que je peux jouer sur une ou deux charges résistives absorber le surplus. En été, les clims ou pompes de piscine commandées en Tout ou Rien permettraient 'd'étaler' le routage sur une période de la journée beaucoup plus ngue.

### 6. Calibration..
> Je vois:
>
> > f_powerCal{ 0.04320F, 0.04374F, 0.04385F };
> > f_voltageCal{ 0.7892F, 0.8002F, 0.7995F };
>
> Faut-il calibrer également la tension ?

La tension doit être calibrée si on veut voir des valeurs exactes. C'est purement informatif, cela n'a aucune incidence sur le fonctionnement du routeur.
Petite question en passant : où as-tu trouvé ces valeurs ? ce sont les valeurs que tu as trouvé sur ton propre routeur ?

### 7. Offset de routage
> Y a t'il un paramètre d'offset pour dire au routeur de rechercher par exemple 20W d'import (ou d'export)

Oui, on peut bidouiller la variable `REQUIRED_EXPORT_IN_WATTS`. La valeur peut être positive ou négative. C'est une valeur assez théorique et approximative. Sachant qu'on travaille en burst, et compte tenu de l'imprécision des pinces, pollution du courant, ... si on veut 20W d'import, il faudra peut-être mettre `-40`.

Voilà voilà.

Fred

**Reply by cvfe13** [reply]
Tout d'abord, merci pour cette réponse rapide...

Point 1.	Sonde de température
	L'ecriture de code en C++ dépasse quelque peu mes connaissances en programmation, mais si on peut accéder à ces valeurs avec EmonESP comme expliqué au point 4, je vais voir si j'arrive à programmer un ESP avec EmonESP.
			
Points 2 & 3.
	Je pensais commander la rotation des priorités ainsi que la marche forcée (Boost) via une carte relais additionnelle.
	J'ai une carte 4 relais + ESP12F que j'ai flashée avec ESPHome, ce qui me permet de piloter les 4 relais via Home Assistant. Ces relais serviront à activer le mode Boost et à commander la rotation.
	
Point 4.	EmonESP
	Je viens de faire des recherches sur EmonESP....
	Finalement EmonESP serait la clé de voute de l'ensemble, si il permet de lire la température, la puissance de routage ainsi que la commande Rotation et Boost.
	D'autant plus que EmonESP utilise MQTT, ce qui est parfait pour communiquer avec Home Assistant !
	
Point 5.	??  Peut-on donner une priorité à une sortie Relais par rapport à la sortie Triac ?

Point 6.	Ces valeurs sont un copier/coller d'un post de yannick68 (https://forum-photovoltaique.fr/viewtopic.php?t=42852&start=260)

Point 7.	REQUIRED_EXPORT_IN_WATTS
	J'avais bien vu ce paramètre dans config_system.h, mais je pensais que c'était spécifique à la notion du profil Zéro Export d'Enphase
	

Comme dit dans mon post précedent, je viens juste de finir le montage, je n'ai pas encore fait la calibration. Ce sera pour le début de semaine...

Francis

**Reply by FredM67** [reply]
> Tout d'abord, merci pour cette réponse rapide...
> 
> Point 1. Sonde de température L'ecriture de code en C++ dépasse quelque peu mes connaissances en programmation, mais si on peut accéder à ces valeurs avec EmonESP comme expliqué au point 4, je vais voir si j'arrive à programmer un ESP avec EmonESP.
> 
> Points 2 & 3. Je pensais commander la rotation des priorités ainsi que la marche forcée (Boost) via une carte relais additionnelle. J'ai une carte 4 relais + ESP12F que j'ai flashée avec ESPHome, ce qui me permet de piloter les 4 relais via Home Assistant. Ces relais serviront à activer le mode Boost et à commander la rotation.

Je réalise fréquemment ceci via une carte multi-relais Wifi. Il y au moins un exemple de routeur dans la galerie photos.

> Point 4. EmonESP Je viens de faire des recherches sur EmonESP.... Finalement EmonESP serait la clé de voute de l'ensemble, si il permet de lire la température, la puissance de routage ainsi que la commande Rotation et Boost. D'autant plus que EmonESP utilise MQTT, ce qui est parfait pour communiquer avec Home Assistant !

Oui EmonESP peut être une bonne solution pour cela.

> Point 5. ?? Peut-on donner une priorité à une sortie Relais par rapport à la sortie Triac ?

C'est compliqué et cela ne donne pas beaucoup de sens. Un relais se déclenche à partir d'un certain surplus. Il faudrait alors attendre que ce surplus soit atteint, et donc le perdre/revendre pour que le 1er relais démarre et ensuite seulement le ou les triacs utiliseraient le surplus du surplus. Quelle serait l'application d'une telle configuration ?

> Point 6. Ces valeurs sont un copier/coller d'un post de yannick68 (https://forum-photovoltaique.fr/viewtopic.php?t=42852&start=260)

Okay :D

> Point 7. REQUIRED_EXPORT_IN_WATTS J'avais bien vu ce paramètre dans config_system.h, mais je pensais que c'était spécifique à la notion du profil Zéro Export d'Enphase

C'est aussi utiliser pour le zéro export. C'est simplement une variable d'offset.

> Comme dit dans mon post précedent, je viens juste de finir le montage, je n'ai pas encore fait la calibration. Ce sera pour le début de semaine...
> 
> Francis

Fred

**Comment by cvfe13** [comment]
Quelques nouvelles de l'avancement....

Aujourd'ui je me suis attelé à la calibration du routeur.
J'ai televersé le sketch Mk2_3phase_RFdatalog_temp.ino puis cal_CTx_v_meter.ino
Je n'ai pas vu de difference dans le moniteur série.
Quelle différence entre ces deux sketchs au niveau de la calibration ?

Voilà deux lignes du moniteur série:
3641.88, P:-2098, P1:2, P2:-1381, P3:-719, V1:234.88, V2:229.35, V3:235.29, (minSampleSets/MC 32, #ofSampleSets 8011)
3640.19, P:-2096, P1:2, P2:-1381, P3:-717, V1:235.00, V2:229.13, V3:235.24, (minSampleSets/MC 32, #ofSampleSets 8011) 
Si les dernières valeurs sont explicites, à quoi correspond la première valeur ?

Pour information, voilà mes valeurs de calibration:
inline constexpr float f_powerCal[NO_OF_PHASES]{ 0.04293F, 0.04246F, 0.04282F };
inline constexpr float f_voltageCal[NO_OF_PHASES]{ 0.8060F, 0.8060F, 0.8062F };

J'ai également flashé un Wemos 8266 D1mini avec EmonESP
Si j'arrive à le connecter au routeur ce serait parfait, car EmonESP affiche les températures des sondes Dallas et la puissance de chaque phase.
En revanche , je n'ai pas trouvé comment on pouvait commuter un ou deux relais pour le Mode Boost et Rotation de priorité dans EmonESP ?
Concernant l'adaptation série du routeur et du Wemos, penses-tu qu'il faut un Level Shifter pour adapter les signaux ?
ou éventuellement un schéma de ce que tu as utilisé ?

Merci, Francis

**Reply by FredM67** [reply]
> Quelques nouvelles de l'avancement....
> 
> Aujourd'ui je me suis attelé à la calibration du routeur. J'ai televersé le sketch Mk2_3phase_RFdatalog_temp.ino puis cal_CTx_v_meter.ino Je n'ai pas vu de difference dans le moniteur série. Quelle différence entre ces deux sketchs au niveau de la calibration ?

Ce qui est affiché est identique. Dans le sketch de calib, il n'y a que le code de mesure. Tout le reste des fonctionnalités est absent.

> Voilà deux lignes du moniteur série: 3641.88, P:-2098, P1:2, P2:-1381, P3:-719, V1:234.88, V2:229.35, V3:235.29, (minSampleSets/MC 32, #ofSampleSets 8011) 3640.19, P:-2096, P1:2, P2:-1381, P3:-717, V1:235.00, V2:229.13, V3:235.24, (minSampleSets/MC 32, #ofSampleSets 8011) Si les dernières valeurs sont explicites, à quoi correspond la première valeur ?

La 1ère valeur est représente le buffer utilisé pour calculer l'énergie à l'instant t. Elle peut servir à débugguer, quand je teste de nouveqx trucs par exemple. Le nb 32 correspond au nombre d'ensemble de mesures réalisées pendant un cycle entier (20 ms). Un ensemble de mesures correspond à 6 mesures (3 tensions; 3 courants). Le dernier nb est le nombre d'ensemble de mesures réalisées entre 2 affichages de log, soit 5 secondes.

> Pour information, voilà mes valeurs de calibration: inline constexpr float f_powerCal[NO_OF_PHASES]{ 0.04293F, 0.04246F, 0.04282F }; inline constexpr float f_voltageCal[NO_OF_PHASES]{ 0.8060F, 0.8060F, 0.8062F };

Les valeurs paraissent correctes. La calibration de la tension n'est pas nécessaire, sauf si on souhaite avoir les valeurs correctes pour l'affichage.
 
> J'ai également flashé un Wemos 8266 D1mini avec EmonESP Si j'arrive à le connecter au routeur ce serait parfait, car EmonESP affiche les températures des sondes Dallas et la puissance de chaque phase. En revanche , je n'ai pas trouvé comment on pouvait commuter un ou deux relais pour le Mode Boost et Rotation de priorité dans EmonESP ? Concernant l'adaptation série du routeur et du Wemos, penses-tu qu'il faut un Level Shifter pour adapter les signaux ? ou éventuellement un schéma de ce que tu as utilisé ?

C'est normal, car le boost via relais n'est pas implémenté. Personne ne me l'a jamais demandé :D Si nécessaire, je peux le rajouter !
Si les 2 ont les mêmes tensions pour les entrées/sorties, on peut brancher directement en veillant à avoir une masse commune.

> Merci, Francis

**Comment by cvfe13** [comment]
Merci pour le decryptage de la ligne du moniteur série.
Aujourd'hui j'ai installé une carte relais, installé deux sondes Dallas et activé EmonESP.
Après activation, les deux sondes sont bien vues par le programme, et j'ai un paramètre supplémentaire qui s'affiche pour la Puissance.
`3671.90, P:-3624/-449, P1:0, P2:-2178, P3:-1446, V1:231.62, V2:232.35, V3:234.55, T1:28.37, T2:28.12, (minSampleSets/MC 32, #ofSampleSets 8016)`

> C'est normal, car le boost via relais n'est pas implémenté. Personne ne me l'a jamais demandé :D Si nécessaire, je peux le rajouter !

Je me suis mal exprimé. Vu que EmonESP communique avec le routeur, je pensais que l'on pouvait commander aussi le mode Boost directement via EmonESP.

> Si les 2 ont les mêmes tensions pour les entrées/sorties, on peut brancher directement en veillant à avoir une masse `commune.

J'ai donc activé EmonESP en conservant le moniteur série:

```
P:-3340,P1:9,P2:-2054,P3:-1295,L1:100.00,L2:100.00,T1:28.62,T2:28.31
P:52,P1:6,P2:286,P3:-240,L1:0.00,L2:0.00,T1:29.50,T2:29.50
```

Il affiche bien les puissances, les températures. Je suppose que L1 et L2 sont les pourcentages de routage des cartes triac ?

Ensuite, j'ai décommenté la ligne 
`#define SERIALOUT /**< Uncomment if a wired serial connection is used */`

puis connécté la liaison série physique
	Wemos	-->	Mk2
	Rx		-->		Tx
	Tx		-->		Rx
	GND		-->		GND
	Vcc		-->		5V du régulateur 7805 installé sur le PCB du routeur pour l'alimentation des sondes Dallas
Les tensions Rx et Tx du Wemos sont bien à 3.3V.

Je ne sais pas s'il faut commenter la ligne
`#define SERIALPRINT  /**< include 'human-friendly' print statement for commissioning - comment this line to exclude. */`
lorsqu'on active la liaison série physique.

Mais rien ne remonte ...
Pourtant j'ai bien une remontée d'infos lorsque je saisis
http://\<adresse de EmonESP\>/input?string=CT1:400,CT2:300,CT3:600,T0:20.5,T1:18.5

**Reply by FredM67** [reply]
> Merci pour le decryptage de la ligne du moniteur série. Aujourd'hui j'ai installé une carte relais, installé deux sondes Dallas et activé EmonESP. Après activation, les deux sondes sont bien vues par le programme, et j'ai un paramètre supplémentaire qui s'affiche pour la Puissance. `3671.90, P:-3624/-449, P1:0, P2:-2178, P3:-1446, V1:231.62, V2:232.35, V3:234.55, T1:28.37, T2:28.12, (minSampleSets/MC 32, #ofSampleSets 8016)`
> 
> > C'est normal, car le boost via relais n'est pas implémenté. Personne ne me l'a jamais demandé :D Si nécessaire, je peux le rajouter !
> 
> Je me suis mal exprimé. Vu que EmonESP communique avec le routeur, je pensais que l'on pouvait commander aussi le mode Boost directement via EmonESP.

Peut-être bien qu'on ne parle pas de la même chose. Si tu as l'affichage ci-dessus, alors tu as activé la fonction délestage par sortie tout-ou-rien (relais). Pour ces sorties, je n'ai pas prévu de boost, car personne ne l'a jamais demandé. Mais si nécessaire, je peux le rajouter.

Avec EmonESP (enfin pas directement EmonESP officiel, il faut une version customisée), il est possible de commander le boost. Dans ma version du prog dérivée de la version officielle EmonESP, les 3 fonctions que j'ai implémentée sont :

- routage ON/OFF avec possibilité de saisir des dates d'arrêt de routage et de début de routage (pratique pour les vacances, pour remettre le routage en route genre la veille du retour de vacances).
- rotation des priorités
- marche forcée. Si mes souvenirs sont bons, cette marche forcée est ou peut être asservie à la température.

> > Si les 2 ont les mêmes tensions pour les entrées/sorties, on peut brancher directement en veillant à avoir une masse `commune.
> 
> J'ai donc activé EmonESP en conservant le moniteur série:
> 
> ```
> P:-3340,P1:9,P2:-2054,P3:-1295,L1:100.00,L2:100.00,T1:28.62,T2:28.31
> P:52,P1:6,P2:286,P3:-240,L1:0.00,L2:0.00,T1:29.50,T2:29.50
> ```
> 
> Il affiche bien les puissances, les températures. Je suppose que L1 et L2 sont les pourcentages de routage des cartes triac ?

C'est tout à fait cela. C'est le taux de routage moyen sur une fenêtre fixe de 5 secondes (intervalle d'affichage du log).

> 
> Ensuite, j'ai décommenté la ligne `#define SERIALOUT /**< Uncomment if a wired serial connection is used */`

Si `EMONESP` est défini, alors `SERIALOUT` sera désactivé (cf début de main.cpp).

> 
> puis connécté la liaison série physique Wemos --> Mk2 Rx --> Tx Tx --> Rx GND --> GND Vcc --> 5V du régulateur 7805 installé sur le PCB du routeur pour l'alimentation des sondes Dallas Les tensions Rx et Tx du Wemos sont bien à 3.3V.
> 
> Je ne sais pas s'il faut commenter la ligne `#define SERIALPRINT /**< include 'human-friendly' print statement for commissioning - comment this line to exclude. */` lorsqu'on active la liaison série physique.

Idem, si `EMONESP` est défini, alors `SERIALPRINT` sera désactivé (cf début de main.cpp).

> 
> Mais rien ne remonte ... Pourtant j'ai bien une remontée d'infos lorsque je saisis http://<adresse de EmonESP>/input?string=CT1:400,CT2:300,CT3:600,T0:20.5,T1:18.5

Bon, dès que j'ai le temps, faut que je re-teste chez moi... peut-être j'ai cassé quelque chose entre temps, ca fait un bail que je n'ai pas testé cette histoire d'EmonESP.

**Reply by FredM67** [reply]
À l'adresse quatre-vingt deux point soixante quatre point cinquante six point cent soixante-dix neuf, sur le port deux mille cinq cent vingt cinq, tu peux voir ce que j'ai fait comme interface.
Tu peux jouer avec, rien n'est branché sur ce module.

**Comment by cvfe13** [comment]
Bonjour Fred
J'ai encore quelques points qui ne sont pas clairs concernant le routeur triphasé
Si le routage vers le chauffe-eau marche bien, il n'en est pas de même pour les deux autres sorties triac.

J'ai activé trois sorties Triac.
```
inline constexpr uint8_t physicalLoadPin[NO_OF_DUMPLOADS]{ 5, 6, 7 };    /**< for 3-phase PCB, Load #1/#2/#3 (Rev 2 PCB) */
inline constexpr uint8_t loadPrioritiesAtStartup[NO_OF_DUMPLOADS]{ 0, 1, 2 }; /**< load priorities and states at startup */
```
La première (pin5) est la resistance monophasée de mon ECS (3000W)
La seconde  (pin6) est un chauffage purement résistif (1500W)
La troisième (pin7) est un également un chauffage purement résistif (1200W)

J'ai activé l'otion rotation de phase 
`inline constexpr uint8_t rotationPin{ 10 };   /**< if LOW, trigger a load priority rotation */ `
Pour cela, j'ai un bouton poussoir pour effectuer la rotation des sorties	

Je n'ai pas activé l'option dual tarif
`inline constexpr bool DUAL_TARIFF{ false };          /**< set it to 'true' if there's a dual tariff each day AND the router is connected to the billing meter */`

J'ai activé EmonESP, pour avoir une info sur l'ouverture du triac dans le serial monitor, mais j'ai abandonné l'idée de gérer le routeur via EmonESP, le projet étant trop ancien et pas maintenu à jour.

Comment est géré le fonctionnement des trois sorties triac ?
sont-elles sequentielles ? par exemple envoi du surplus dans l'ECS jusqu'à ce que le thermostat coupe, puis envoi vers la 2ème sortie, puis vers la troisième.
ou sont elles simultanées dans les 3 charges ?

Si j'ai 4000W de surplus va t'il envoyer 3000 W dans l'ECS et 1000 W dans le 1er radiateur ?

Pour l'instant, le routage sur la sortie Triac 1 fonctionne correctement, mais j'ai un fonctionnement aléatoire des voies 2 et 3. Parfois même le Triac 2 est ouvert à 100% sans production PV !

La rotation des priorités n'a pas l'air de fonctionner également. Un appui sur le poussoir n'a aucun effet.
  

J'ai activé le mode Boost
`inline constexpr uint8_t forcePin{ 11 };      /**< for 3-phase PCB, force pin */`
et cela fonctionne.
Qu'en est-il si on active le mode Boost et qu'on est en priorité 2,3,1 ? Le mode Boost va t'il forcer le chauffage 1 sur maximum ?
Je voudrais que le mode Boost ne force que la voie 1 (càd l'ECS) uniquement pour faire un complément d'eau chaude en hiver lorsqu'il n'y a pas eu assez de soleil.

Merci d'eclairer ma lanterne ...

**Reply by FredM67** [reply]
> Bonjour Fred J'ai encore quelques points qui ne sont pas clairs concernant le routeur triphasé Si le routage vers le chauffe-eau marche bien, il n'en est pas de même pour les deux autres sorties triac.
> 
> J'ai activé trois sorties Triac.
> 
> ```
> inline constexpr uint8_t physicalLoadPin[NO_OF_DUMPLOADS]{ 5, 6, 7 };    /**< for 3-phase PCB, Load #1/#2/#3 (Rev 2 PCB) */
> inline constexpr uint8_t loadPrioritiesAtStartup[NO_OF_DUMPLOADS]{ 0, 1, 2 }; /**< load priorities and states at startup */
> ```
> 
> La première (pin5) est la resistance monophasée de mon ECS (3000W) La seconde (pin6) est un chauffage purement résistif (1500W) La troisième (pin7) est un également un chauffage purement résistif (1200W)
> 
> J'ai activé l'otion rotation de phase `inline constexpr uint8_t rotationPin{ 10 }; /**< if LOW, trigger a load priority rotation */ ` Pour cela, j'ai un bouton poussoir pour effectuer la rotation des sorties
> 
> Je n'ai pas activé l'option dual tarif `inline constexpr bool DUAL_TARIFF{ false }; /**< set it to 'true' if there's a dual tariff each day AND the router is connected to the billing meter */`
> 
> J'ai activé EmonESP, pour avoir une info sur l'ouverture du triac dans le serial monitor, mais j'ai abandonné l'idée de gérer le routeur via EmonESP, le projet étant trop ancien et pas maintenu à jour.
Que le projet ne soit plus maintenu n'est pas un problème en soit. Tant que le prog fait ce dont on a besoin , tout baigne.
> 
> Comment est géré le fonctionnement des trois sorties triac ? sont-elles sequentielles ? par exemple envoi du surplus dans l'ECS jusqu'à ce que le thermostat coupe, puis envoi vers la 2ème sortie, puis vers la troisième. ou sont elles simultanées dans les 3 charges ?

Oui, c'est séquentiel par défaut.
> 
> Si j'ai 4000W de surplus va t'il envoyer 3000 W dans l'ECS et 1000 W dans le 1er radiateur ?

Oui, c'est exactement cela.
> 
> Pour l'instant, le routage sur la sortie Triac 1 fonctionne correctement, mais j'ai un fonctionnement aléatoire des voies 2 et 3. Parfois même le Triac 2 est ouvert à 100% sans production PV !

C'est très louche cette affaire !
> 
> La rotation des priorités n'a pas l'air de fonctionner également. Un appui sur le poussoir n'a aucun effet.

As-tu aussi changer la ligne :
`inline constexpr RotationModes PRIORITY_ROTATION{ RotationModes::PIN };`
Le 8 août, j'ai trouvé un bug au niveau de cette fonctionnalité, peut-être as-tu encore la version non corrigée?
> 
> J'ai activé le mode Boost `inline constexpr uint8_t forcePin{ 11 }; /**< for 3-phase PCB, force pin */` et cela fonctionne. Qu'en est-il si on active le mode Boost et qu'on est en priorité 2,3,1 ? Le mode Boost va t'il forcer le chauffage 1 sur maximum ? Je voudrais que le mode Boost ne force que la voie 1 (càd l'ECS) uniquement pour faire un complément d'eau chaude en hiver lorsqu'il n'y a pas eu assez de soleil.

Par défaut, le boost active toutes les sorties triacs, peu importe les priorités.
Si on veut activer seulement certaines sorties, il faut modifier le code. Je n'ai pas encore eu le temps de faire un système de configuration pour cela...
Je te modifie le code pour booster uniquement 1... je te fais signe quand j'ai fait la modif.

> 
> Merci d'eclairer ma lanterne ...

**Comment by cvfe13** [comment]
`Le 8 août, j'ai trouvé un bug au niveau de cette fonctionnalité, peut-être as-tu encore la version non corrigée?`

Effectivement, j'avais installé la version téléchargée au mois de juillet...
Et la modification a résolu mes problèmes ...

J'ai de la chance, car le soleil vient de pointer son nez et j'ai pu effectuer quelques tests.
La rotation de phase marche mainten,ant comme prévu, et le transfert des charges s'effectue au fur et à mesure que le triac arrive à 100%

Encore merci pour ton aide.

![routage](https://github.com/user-attachments/assets/b399d127-1d87-4945-b66b-555210a933c2)
![routeur1](https://github.com/user-attachments/assets/f67cd5c5-0633-481e-a077-9882b2663266)

**Reply by FredM67** [reply]
> `Le 8 août, j'ai trouvé un bug au niveau de cette fonctionnalité, peut-être as-tu encore la version non corrigée?`
> 
> Effectivement, j'avais installé la version téléchargée au mois de juillet... Et la modification a résolu mes problèmes ...

Si un jour tu trouves un autre bug, tu peux écrire une "Issue"... Il n'y a aucun mal à écrire une Issue qui au final n'est pas forcément un bug mais peut-être une erreur de config ou autre.
Mais au moins, si c'est un bug, j'ai une trace ;-)

> J'ai de la chance, car le soleil vient de pointer son nez et j'ai pu effectuer quelques tests. La rotation de phase marche maintenant comme prévu, et le transfert des charges s'effectue au fur et à mesure que le triac arrive à 100%
> 
> Encore merci pour ton aide.

De rien ;-)

**Comment by cvfe13** [comment]
Bonjour Fred
Je reviens vers toi car j'ai toujours mon problème d'override des charges.
J'ai 3 charges connectées au routeur (ECS et 2 radiateurs), mais j'aimerai ne forcer que l'ECS lorsque j'active le mode Override.

J'ai retrouvé un de tes posts ici : https://forum-photovoltaique.fr/viewtopic.php?t=42852&start=240
où tu disais qu'il suffit de modifier la fonction forceFullPower ainsi :

```
Remplacer:
	for (auto &bForceLoad : b_forceLoadOn)
		bForceLoad = !pinState;
par
	b_forceLoadOn[0] = !pinState;

```
mais ces variables ont été remplacées. Comment dois-je procéder pour ne forcer que l'ECS


D'autre part, je voulais récuperer l'ordre de priorité de routage car j'utilise les données envoyées sur la liaison série que je récupère dans Home Assistant.
Pour cela j'ai modifié la fonction printForEmonESP(const bool bOffPeak) du utils.h en ajoutant les lignes suivantes:
```
...
// affiche Rotation en cours
  Serial.print(F(",R:"));
  for (const auto& loadPrioAndState : loadPrioritiesAndState)
  {
    Serial.print(loadPrioAndState);
  }
...
```
mais souvent j'ai des affichages erratiques... par ex affichage à 6 ou 9 chiffres ou plus. Aurais-tu une methode plus sûre ?
Merci

**Reply by FredM67** [reply]
Salut,
Oui, depuis énormément de choses ont été modifiées, dans le tri comme dans le mono. J'ai même fait un module ESPHome-Home Assistant pour le Mk2 :D En gros, tes vues Home Assistant sont maintenant faisables en quelques secondes.
Au niveau du prog du routeur, j'ai aussi rajouté une sortie série pour Home Assistant. Je me suis basé sur la TIC du Linky en utilisant des tags, ... Ainsi, pas ou très peu de problèmes de corruption de données, car il y a un checksum pour chaque couple "tag-valeur".
[Doc Mk2-ESPHome](https://gist.github.com/FredM67/986e1cb0fc020fa6324ccc151006af99)

Pour forcer uniquement la charge 1, c'est lignes 80-83 dans main.cpp, comme ceci:
```cpp
    // for (auto &bOverrideLoad : Shared::b_overrideLoadOn)
    // {
    //   bOverrideLoad = !pinState;
    // }
    Shared::bOverrideLoad[0] = !pinState;
```

Pour retourner l'ordre des charges dans Home Assistant, pour l'instant je n'ai rien prévu, mais oui, on peut le faire. Reste à savoir comment. L'envoyer toutes les n secondes n'a pas de sens, ca encombre la sortie série pour rien. Le renvoyer à chaque changement est une solution mais si tu "rates" le message... ben tu perds l'info !

**Comment by cvfe13** [comment]
Merci, mais j'ai eu une erreur à la compilation...
```
C:\Users\bmart\Documents\Arduino\Mk2_3phase_RFdatalog_temp\main.cpp: In function 'bool forceFullPower()':
C:\Users\bmart\Documents\Arduino\Mk2_3phase_RFdatalog_temp\main.cpp:157:5: error: 'bOverrideLoad' was not declared in this scope
     bOverrideLoad[0] = !pinState;
     ^~~~~~~~~~~~~
C:\Users\bmart\Documents\Arduino\Mk2_3phase_RFdatalog_temp\main.cpp:157:5: note: suggested alternative: 'b_overrideLoadOn'
     bOverrideLoad[0] = !pinState;
     ^~~~~~~~~~~~~
     b_overrideLoadOn

exit status 1

Compilation error: 'bOverrideLoad' was not declared in this scope
```
J'ai donc remplacé bOverrideLoad[0] = !pinState;
par  b_overrideLoadOn[0] = !pinState;
et j'ai pu compiler le sketch....

Effectivement, ton script ESPHome m'aurait bien facilité le travail l'an dernier quand j'ai installé le routeur. Je vais voir comment migrer pour installer ta dernière version du nano du Mk2.

**Reply by FredM67** [reply]
> Merci, mais j'ai eu une erreur à la compilation...
> 
> ```
> C:\Users\bmart\Documents\Arduino\Mk2_3phase_RFdatalog_temp\main.cpp: In function 'bool forceFullPower()':
> C:\Users\bmart\Documents\Arduino\Mk2_3phase_RFdatalog_temp\main.cpp:157:5: error: 'bOverrideLoad' was not declared in this scope
>      bOverrideLoad[0] = !pinState;
>      ^~~~~~~~~~~~~
> C:\Users\bmart\Documents\Arduino\Mk2_3phase_RFdatalog_temp\main.cpp:157:5: note: suggested alternative: 'b_overrideLoadOn'
>      bOverrideLoad[0] = !pinState;
>      ^~~~~~~~~~~~~
>      b_overrideLoadOn
> 
> exit status 1
> 
> Compilation error: 'bOverrideLoad' was not declared in this scope
> ```
> 
> J'ai donc remplacé bOverrideLoad[0] = !pinState; par b_overrideLoadOn[0] = !pinState; et j'ai pu compiler le sketch....
> 
> Effectivement, ton script ESPHome m'aurait bien facilité le travail l'an dernier quand j'ai installé le routeur. Je vais voir comment migrer pour installer ta dernière version du nano du Mk2.

désolé, c'est ```Shared::b_overrideLoadOn[0]``` ! J'ai du faire un copier-coller au lieu de retaper à la main !
